### PR TITLE
add support for `intersection` & `union` in search operations

### PR DESCRIPTION
**Pull request by dsa0x**
#### Summary

This PR addresses this enhancement issue https://github.com/sigstore/rekor/issues/849, which suggested adding support for intersection and union search. The PR adds support for `and` and `or` operators.
- `and` operation returns a union of values in each list (i.e no duplicates)
- `or` returns a list of all values, not minding duplicates.
#### Release Note

new:   support for intersection and union search through the use of `and` and `or` operators.

### CODE REVIEW

**Comment by dlorenc** [comment]
Oh wow, this is really cool!

File: cmd/rekor-cli/app/search.go [hunk 1]
```diff
@@ -58,6 +58,8 @@ func addSearchPFlags(cmd *cobra.Command) error {
 	cmd.Flags().Var(NewFlagValue(shaFlag, ""), "sha", "the SHA256 or SHA1 sum of the artifact")
 
 	cmd.Flags().Var(NewFlagValue(emailFlag, ""), "email", "email associated with the public key's subject")
+
+	cmd.Flags().Var(NewFlagValue(operatorFlag, ""), "operator", "operator to use for the search")
```
Comment (asraa) [inline cmd/rekor-cli/app/search.go]: nit: add supported value are `and` and `or`

File: pkg/api/index.go [hunk 2]
```diff
@@ -100,3 +107,32 @@ func addToIndex(ctx context.Context, key, value string) error {
...
 }
+
+type Container interface {
+	Add([]string)
+	Result() []string
+}
+
+type Uniq map[string]struct{}
+
+func (u Uniq) Add(elements []string) {
```
Comment (asraa) [inline pkg/api/index.go]: Nice implementation!

Question: If this is for `And` then won't we want an intersection? This seems to add any one of them, with a unique result.

For example, I expect the last query to be the intersection of the first two results, and only have 1 UUID
```
asraa@asraa1:~/git/rekor$ ./rekor search --public-key publickey.pem  --pki-format x509 --rekor_server http://localhost:3000
Found matching entries (listed by UUID):
64b1e670a1cf84b4d636d2d578fea277126279529759f537879796d8480f284f05708ba02b10ca38
64b1e670a1cf84b4af3fda3822b3db033f55be3af22324fa16381dc432d60d1ef1bd7ea7036538b1
asraa@asraa1:~/git/rekor$ ./rekor search --artifact COPYRIGHT.txt --rekor_server http://localhost:3000
Found matching entries (listed by UUID):
64b1e670a1cf84b47d1d6bedeb657dde5830a09aea72c528358786ee37183d7fb547f05692bb8af0
64b1e670a1cf84b4d636d2d578fea277126279529759f537879796d8480f284f05708ba02b10ca38
asraa@asraa1:~/git/rekor$ ./rekor search --public-key publickey.pem --operator and --artifact COPYRIGHT.txt  --pki-format x509 --rekor_server http://localhost:3000
Found matching entries (listed by UUID):
64b1e670a1cf84b47d1d6bedeb657dde5830a09aea72c528358786ee37183d7fb547f05692bb8af0
64b1e670a1cf84b4d636d2d578fea277126279529759f537879796d8480f284f05708ba02b10ca38
64b1e670a1cf84b4af3fda3822b3db033f55be3af22324fa16381dc432d60d1ef1bd7ea7036538b1
```

**Review by asraa — COMMENTED** [review]
Thanks! Quick question on correctness/expectation here.

File: pkg/api/index.go (hunk 2 above)
Comment (dsa0x) [reply pkg/api/index.go]: I think intersection/union are not the correct words to use here. It's not a set. Basically, what we are returning is either the union of both values (i.e no duplicates), or just return all values, not minding the duplicates.

I am not sure if there is usecase for intersection, or we could have an operator like `intersect` for intersection. what do you think?

Comment (asraa) [reply pkg/api/index.go]: What is the use-case of not minding duplicates? For anyone iterating through results, that would be inefficient.

Yes! I thought intersection is the most useful: e.g. "I want signatures on package X signed by user Y."

Comment (dsa0x) [reply pkg/api/index.go]: Yeah. That makes sense. I'll go about it this way. if there's no operator, it'll work like it does now, and return all values. Then we can use the operator for intersection or union.

File: cmd/rekor-cli/app/search.go (hunk 1 above)
Comment (dsa0x) [reply cmd/rekor-cli/app/search.go]: addressed in 12b38af93af0481aaf44015a3cd4462bfc48d486

**Review by asraa — APPROVED** [review]
This looks amazing now! Great work. Good job on the unit tests and testing in e2e. 
The implementation was very cool to read.

**Review by dlorenc — APPROVED** [review]

**Pull request merged**